### PR TITLE
Implement Route class

### PR DESCRIPTION
**Pull request by joshyuan360**
Create HTTP Route class that specifies a request type and a path, and provides functionality for determining whether or not two paths match. Invalid paths will throw exception.

### CODE REVIEW

File: express/express/route.cc [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#include "route.h"
+#include <assert.h>
```
Comment (mysterytony) [inline express/express/route.cc]: take out

File: express/express/route.h [hunk 2]
```diff
@@ -1,24 +1,29 @@
...
-	template <>
-	class hash<CExpress::Route>
-	{
-	public:
-		size_t operator() (const CExpress::Route& r) const;
-	};
-}
 
 namespace CExpress {
+    enum RequestType { POST, GET };
```
Comment (mysterytony) [inline express/express/route.h]: indent this and the class correctly

File: express/express/route.cc [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+#include <iostream>
+
+using namespace CExpress;
+
+Route::Route(RequestType t, std::string p) {
+    if (!Route::validPath(p)) {
+        throw "Input path invalid."; // todo: use exception
```
Comment (mysterytony) [inline express/express/route.cc]: you hvae a TODO?!

File: express/express/route.cc [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    }
+    
+    type = t;
+    path = Route::reducePath(p);
+}
+
+bool Route::validPath(std::string input) { // todo: make validation more strict
+    if (input.empty() || input[0] != '/' || input[input.size() - 1] == '/')
```
Comment (mysterytony) [inline express/express/route.cc]: put comment here

File: express/express/route.cc [hunk 5]
```diff
@@ -0,0 +1,40 @@
+#include "route.h"
```
Comment (mysterytony) [inline express/express/route.cc]: rename file to cpp

File: express/express/route.h [hunk 6]
```diff
@@ -1,24 +1,29 @@
 #pragma once
 
 #include <functional>
+#include <vector>
+#include <string>
 
 namespace CExpress {
```
Comment (mysterytony) [inline express/express/route.h]: remove namespace

File: express/express/route.h (hunk 2 above)
Comment (mysterytony) [inline express/express/route.h]: there more request types

**Comment by mysterytony** [comment]
Also need a Route comparison function. @joshyuan360

File: express/express/route.h [hunk 7]
```diff
@@ -1,24 +1,24 @@
...
-}
+#include <string>
 
 namespace CExpress {
+    enum RequestType { GET, HEAD, POST, PUT, DELETE, CONNECT, OPTIONS, TRACE, PATCH };
 	class Route {
-		friend size_t std::hash<Route>::operator() (const Route&) const;
+        public:
+            /* Create a Route with the specified request type and path. */
+            Route(RequestType type, std::string path);
```
Comment (mysterytony) [inline express/express/route.h]: constructor should take 2 strings

File: express/express/route.h [hunk 8]
```diff
@@ -1,24 +1,24 @@
...
+            /* Get the path of this Route. */
+            std::string getPath();
+            /* Get the request type of this Route. */
+            RequestType getRequestType();
+            /* True if source and target routes have the same type and path. */
+            bool equals(Route target);
```
Comment (mysterytony) [inline express/express/route.h]: rename this function. as I am checking if a request url is matching the route, i am not comparing two route

**Comment by Xadoy** [comment]
I am waiting for the merging. Once it's done, I will change the code style to fulfill the code style guide

**Comment by mysterytony** [comment]
@Xadoy You don't need to wait for the merge. Just direct edit on his branch.

File: express/express/route.h [hunk 9]
```diff
@@ -1,24 +1,43 @@
...
 
 namespace CExpress {
+    enum RequestType { GET, HEAD, POST, PUT, DELETE, CONNECT, OPTIONS, TRACE, PATCH };
 	class Route {
-		friend size_t std::hash<Route>::operator() (const Route&) const;
+        public:
+            /* Create a Route with the specified request type and path. */
+            Route(RequestType type, std::string path);
+            /* Get the path of this Route. */
+            std::string getPath();
```
Comment (mysterytony) [inline express/express/route.h]: add `const` to all function that will not change the property

File: express/express/route.h [hunk 10]
```diff
@@ -1,24 +1,43 @@
...
+            /* Get the request type of this Route. */
+            RequestType getRequestType();
+            /* True if source and target routes match. */
+            bool matches(Route target);
```
Comment (mysterytony) [inline express/express/route.h]: take reference

File: express/express/route.h [hunk 11]
```diff
@@ -1,24 +1,43 @@
...
+        private:
+            RequestType type;
+            std::string path;
+            std::string reducePath(std::string original);
+            bool validPath(std::string input);
+            std::vector<std::string> split(const std::string &s, char delim);
+            
+            template<typename Out>
+            void split(const std::string &s, char delim, Out result) {
```
Comment (mysterytony) [inline express/express/route.h]: why put function body in `.h` file?

File: express/express/route.cpp [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+bool Route::matches(Route target) {
+    if (getRequestType() != target.getRequestType()) {
+        return false;
+    }
+
+    std::string sourcePath = getPath();
+    std::string targetPath = target.getPath();
+
+    std::vector<std::string> sourceTokens = splitString(sourcePath, '/');
```
Comment (mysterytony) [inline express/express/route.cpp]: directly store the splited vector of this route, dont call split string on this. split in constructor

File: express/express/route.h [hunk 13]
```diff
@@ -1,24 +1,43 @@
...
+                std::stringstream ss;
+                ss.str(s);
+                std::string item;
+                while (std::getline(ss, item, delim)) {
+                    *(result++) = item;
+                }
+            }
+
+            std::vector<std::string> splitString(const std::string &s, char delim) {
```
Comment (mysterytony) [inline express/express/route.h]: pass in reference instead of return a vector

File: express/express/route.cpp [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    std::vector<std::string> targetTokens = splitString(targetPath, '/');
+
+    if (sourceTokens.size() != targetTokens.size()) {
+        return false;
+    }
+
+    for (int i = 0; i < (int)sourceTokens.size(); ++i) {
+        if ((sourceTokens[i].find_first_of(':') == 0) != (targetTokens[i].find_first_of(':') == 0) ||
```
Comment (mysterytony) [inline express/express/route.cpp]: dont use find_first_of. directly store ":" in the vector, so you only need to check the first char

File: express/express/UnitTest/routetest.cpp [hunk 15]
```diff
@@ -0,0 +1,16 @@
+#include "../route.cpp"
+#include <assert.h>
+#include <iostream>
+
+int main() {
+    Route r1(CExpress::POST, "/user/:user/age/:age");
+    Route r2(CExpress::POST, "/user/josh/age/19");
+    Route r3(CExpress::POST, "/user/josh/age/19/");
```
Comment (mysterytony) [inline express/express/UnitTest/routetest.cpp]: add more test cases & corner cases. for example, a single slash. a string that does not begin with a slash. very long string etc

File: express/express/UnitTest/routetest.cpp [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+    
+    assert(r1.matches(r2)); 
+    assert(r2.matches(r3));
+    assert(r1.matches(r3));
+    assert(r1.matches(r1));
+
+    std::cout << "route.cpp unit test completed." << std::endl;
```
Comment (mysterytony) [inline express/express/UnitTest/routetest.cpp]: put router test cases in a function, and call the function in main

File: express/express/UnitTest/routetest.cpp [hunk 17]
```diff
@@ -0,0 +1,16 @@
+#include "../route.cpp"
```
Comment (mysterytony) [inline express/express/UnitTest/routetest.cpp]: test with visual studio, make sure it compiles in VS

File: express/express/route.h (hunk 9 above)
Comment (joshyuan360) [inline express/express/route.h]: const has no effect for return methods

File: express/example/main.cpp [hunk 18]
```diff
@@ -1,5 +1,44 @@
...
+	assert(r1.matches(r1));
+
+	Route r4(CExpress::PATCH, "/name/:name");
+	Route r5(CExpress::GET, "/name/:name");
+	Route r6(CExpress::GET, "/age/:age");
+
+	assert(!r4.matches(r5));
+	assert(!r5.matches(r6));
+
+	std::vector<std::string> exceptionTests = { "/", "", "user/:user" };
```
Comment (mysterytony) [inline express/example/main.cpp]: a single slash should be allowed, in fact, it should be treated same with `/index` `/index.htm` `/index.html` or `` the empty string.

File: express/express/route.cpp [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    
+    std::string sourcePath = getPath();
+    std::string targetPath = target.getPath();
+        
+    std::vector<std::string> sourceTokens;    
+    std::vector<std::string> targetTokens;
+    
+    splitString(sourcePath, '/', targetTokens);	
+    splitString(targetPath, '/', sourceTokens);
```
Comment (mysterytony) [inline express/express/route.cpp]: source tokens should be saved in constructor, no need to construct here

File: express/express/route.cpp [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+
+    if (sourceTokens.size() != targetTokens.size()) {
+        return false;
+    }
+    
+    for (int i = 0; i < (int)sourceTokens.size(); ++i) {
+        if ((sourceTokens[i].find_first_of(':') == 0) != (targetTokens[i].find_first_of(':') == 0)) {
```
Comment (mysterytony) [inline express/express/route.cpp]: dont use find_first_of

directly store ":" so only compare the first char is enough

assume ":" can only exist in source tokens

File: express/express/route.cpp [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+void Route::split(const std::string &s, const char delim, Out result) const {
+	std::stringstream ss;
+	ss.str(s);
+	std::string item;
+	while (std::getline(ss, item, delim)) {
+		*(result++) = item;
+	}
+}
+
+std::vector<std::string> Route::splitString(const std::string &s, const char delim, std::vector<std::string>& elems) const {
```
Comment (mysterytony) [inline express/express/route.cpp]: should return void?

File: express/express/route.cpp [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+        return false;
+    }
+    
+    std::string sourcePath = getPath();
+    std::string targetPath = target.getPath();
+        
+    std::vector<std::string> sourceTokens;    
+    std::vector<std::string> targetTokens;
+    
+    splitString(sourcePath, '/', targetTokens);
```
Comment (mysterytony) [inline express/express/route.cpp]: put delim as a const char in header

File: express/express/route.cpp [hunk 23]
```diff
@@ -0,0 +1,61 @@
+#include "route.h"
+#include <iostream>
+
+using namespace CExpress;
+
+Route::Route(RequestType t, std::string p) {
+    if (p.empty() || p.size() == 1 || p[0] != '/') {
```
Comment (mysterytony) [inline express/express/route.cpp]: no hardcode char, use constant

File: express/express/route.cpp [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+        }
+    }
+    
+    return true;
+}
+
+template<typename Out>
+void Route::split(const std::string &s, const char delim, Out result) const {
+	std::stringstream ss;
+	ss.str(s);
```
Comment (mysterytony) [inline express/express/route.cpp]: put s in ss constructor: `std::stringstream ss(s);`

File: express/express/route.cpp (hunk 20 above)
Comment (joshyuan360) [inline express/express/route.cpp]: going to check both source in target in case target.matches(sources) is called

File: express/express/route.cpp (hunk 24 above)
Comment (joshyuan360) [inline express/express/route.cpp]: doing so throws an exception when creating route with input "//" (which is invalid, but should throw the "invalid path" exception)

File: express/example/main.cpp [hunk 25]
```diff
@@ -1,5 +1,65 @@
...
+	assert(r8.matches(r8));
+
+	Route r10(CExpress::GET, "/user/");
+	Route r11(CExpress::GET, "/user/index.php");
+
+	assert(r10.matches(r11));
+
+	Route r12(CExpress::CONNECT, "/case/:case/");
+	Route r13(CExpress::CONNECT, "/CASE/CASE");
+	Route r14(CExpress::CONNECT, "/case/cAsE/index.cshtml");
```
Comment (mysterytony) [inline express/example/main.cpp]: I think `/index.html` `/index.html` or similar should be left as it is. However, only `/index` and `/` should be the same.

We will use fallback to find the file if the url is `/index.html`

File: express/express/response.h [hunk 26]
```diff
@@ -2,13 +2,29 @@
 #include <unordered_map>
```
Comment (mysterytony) [inline express/express/response.h]: @Xadoy @joshyuan360 why is this file changed?

File: express/express/request.h [hunk 27]
```diff
@@ -3,13 +3,36 @@
 #include <unordered_map>
```
Comment (mysterytony) [inline express/express/request.h]: @Xadoy @joshyuan360 why is this file changed?

**Comment by mysterytony** [comment]
@joshyuan360 Also need to consider the case of query parameter in the end. For example, `/share` and `/share?id=14423` should be the same url.
